File: Diagrammatize.Design.Pattern/code/python/Visitor/A2/SizeVisitor.py

from Visitor import Visitor

class SizeVisitor(Visitor):
    def __init__(self):
        self.size = 0

    def getSize(self):
        return self.size

    def visitFile(self, file_):
        self.size += file_.getSize()

    def visitDirectory(self, directory):
        for entry in directory:
            entry.accept(self)

File: Diagrammatize.Design.Patterns/code/cxx/Visitor/A2/SizeVisitor.cpp
#include "SizeVisitor.hpp"

SizeVisitor::SizeVisitor() {
}

SizeVisitor::~SizeVisitor() {
}

void SizeVisitor::visit(File* file) {
    size += file->getSize();
}

void SizeVisitor::visit(Directory* directory) {
    for (auto entry : *directory) {
        entry->accept(this);
    }
}

int SizeVisitor::getSize() {
    return size;
}
